File: rail-fence-cipher/rail_fence_cipher.rb
class RailFenceCipher

  VERSION = 1

  def initialize(num_rails)
    @num = num_rails || 1

    @rails = Array.new(@num) { |idx| [] }
    @cycle = (0..(@num - 1)).to_a + (1..(@num-2)).to_a.reverse

    @index = 0
  end

  def self.encode(plaintext, num_rails)
    RailFenceCipher.new(num_rails).encode(plaintext)
  end

  def self.decode(ciphertext, num_rails)
    RailFenceCipher.new(num_rails).decode(ciphertext)
  end

  def encode(plaintext)
    build_fence(plaintext.chars).flatten.join
  end

  def decode(ciphertext)
    range = (0..(ciphertext.length - 1))
    fence = build_fence(range.to_a).flatten

    range.map { |num| ciphertext[fence.index(num)] }.join
  end

  private

  def next_index
    idx = @index % @cycle.length
    @index += 1

    @cycle[idx]
  end

  def build_fence(input)
    input.each_with_object(@rails) { |c, rails| rails[next_index] << c }
  end

end

File: elixir/rail-fence-cipher/rail_fence_cipher.exs
defmodule RailFenceCipher do
  @doc """
  Encode a given plaintext to the corresponding rail fence ciphertext
  """
  @spec encode(String.t, pos_integer) :: String.t
  def encode(str, rails) do
    str
    |> String.graphemes
    |> build_fence(rails)
    |> Enum.join
  end

  defp build_fence(chars, rails) do
    chars
    |> Enum.zip(cycle(rails))
    |> Enum.group_by(&group_key/1, &group_val/1)
    |> Map.values
  end

  defp cycle(1), do: Stream.cycle([0])
  defp cycle(rails) do
    0..rails-1
    |> Enum.to_list
    |> Enum.concat(Enum.to_list(rails-2..1))
    |> Stream.cycle
  end

  defp group_key({_letter, rail}), do: rail
  defp group_val({letter, _rail}), do: letter

  @doc """
  Decode a given rail fence ciphertext to the corresponding plaintext
  """
  @spec decode(String.t, pos_integer) :: String.t
  def decode(str, rails) do
    0..(String.length(str)-1)
    |> Enum.to_list
    |> build_fence(rails)
    |> List.flatten
    |> do_decode(str)
  end

  defp do_decode(fence, str) do
    0..(String.length(str)-1)
    |> Enum.map(fn(idx) -> String.at(str, find_spot_in_fence(fence, idx)) end)
    |> Enum.join
  end

  defp find_spot_in_fence(fence, num) do
    Enum.find_index(fence, fn(val) -> val == num end)
  end

end
